lib/whitehall_importer/change_history.rb

```diff
@@ -0,0 +1,41 @@
+module WhitehallImporter
+  class ChangeHistory
+    attr_reader :document_change_history
+
+    def initialize(whitehall_export)
+      @document_change_history = whitehall_document_change_history(whitehall_export).compact
+    end
+
+    def for(edition_number)
+      change_history = document_change_history.select do |edition_change_history|
+        edition_change_history["edition_number"] < edition_number
+      end
+
+      change_history.map { |change| change["change_history_entry"] }.reverse
+    end
+
+  private
+
+    def whitehall_document_change_history(whitehall_export)
+      whitehall_export["editions"].map.with_index do |edition, index|
+        publish_event = EditionHistory.new(edition["revision_history"]).last_state_event("published")
+        edition_number = index + 1
+
+        next if edition["minor_change"] || edition_number == 1
+        next if %w[submitted rejected draft].include?(edition["state"]) && edition["unpublishing"].blank?
+
+        raise AbortImportError, "Edition has a major change but no change note" if edition["change_note"].blank?
+        raise AbortImportError, "Edition has a major change but no publish event" if publish_event.blank?
+
+        {
+          "edition_number" => edition_number,
+          "change_history_entry" => {
+            "id" => SecureRandom.uuid,
+            "note" => edition["change_note"],
+            "public_timestamp" => publish_event["created_at"],
+          },
+        }
+      end
+    end
+  end
```

Comment:
This class looks like a nice encapsulation of the change history concerns.

There's a few concerns I have that there is scatterings of logic slipping into different methods perhaps unnecessarily. For example: the caller of whitehall_document_change_history needs to know that some of the entries are nil and call compact; the for method needs to know that the data is a different order.

I'd wonder if you want to aim for something a bit more compartmentalised. For example (with some variable name tweaks)

```
class ChangeHistory
  attr_reader :entries
  
  def initialize(whitehall_export)
    @entries = build_entries(whitehall_export["editions"])
  end
  
  def for(edition_number)
    entries.each_with_object([]) do |item, memo|
      memo << item["change_history_entry"] if item["edition_number"] < edition_number
    end
  end
  
private

  def build_entries(editions)
    # deals with the looping, reversing and compacting. Calls #history_entry for each item
  end

  def history_entry(edition, edition_number)
    # deals with guard clauses, exceptions and returning the hash
  end
end
```

Reply:
Yep I agree with your concerns & thanks for the comment, I was struggling to brake the class up and your suggestion separates things nicely!! :)